File: src/com/charter/enterprise/collatz/BasicMultithreadedMain.java
package com.charter.enterprise.collatz;

import com.charter.enterprise.collatz.model.CollatzDataModel;
import com.charter.enterprise.collatz.util.Collatz;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

public class BasicMultithreadedMain {
    public static void main(String[] args) throws Exception {
        long startTime = System.currentTimeMillis();
        long endTime = 0L;
        long elapsedTime = 0L;

        List<Future<CollatzDataModel>> collatzFutureList = new ArrayList<>();
        List<CollatzDataModel> collatzList = new ArrayList<>();
        ExecutorService executors = Executors.newFixedThreadPool(3);

        for(int i = 1; i <= 999; i++) {
            collatzFutureList.add(executors.submit(new Collatz(i)));
        }

        for(Future collatzFuture : collatzFutureList) {
            Object o = collatzFuture.get();

            if(o instanceof CollatzDataModel) {
                collatzList.add((CollatzDataModel) o);
            }
        }

        endTime = System.currentTimeMillis();
        elapsedTime = endTime - startTime;

        executors.shutdown();

//        new CollatzInfoPrinter().prettyPrintCollatzList(collatzList);

        System.out.format("the elapsed calculation time is %8d ms", elapsedTime);
    }
}

File: src/basic-multithreaded-main.ts
import { Worker } from 'worker_threads';
import { CollatzDataModel } from './model/collatz-data-model';
import { CollatzInfoPrinter } from './util/collatz-info-printer';
import * as path from 'path';

const THREAD_COUNT = 3;

async function runCollatzInWorker(number: number): Promise<CollatzDataModel> {
    return new Promise((resolve, reject) => {
        const worker = new Worker(path.join(__dirname, 'worker/collatz-worker.js'), {
            workerData: { number }
        });

        worker.on('message', (result) => {
            resolve(new CollatzDataModel(result.collatzNumber, result.collatzIterations));
        });

        worker.on('error', reject);
        worker.on('exit', (code) => {
            if (code !== 0) {
                reject(new Error(`Worker stopped with exit code ${code}`));
            }
        });
    });
}

async function main() {
    const startTime = Date.now();

    const promises: Promise<CollatzDataModel>[] = [];
    const collatzList: CollatzDataModel[] = [];

    // Create a pool of workers
    const tasks: number[] = [];
    for (let i = 1; i <= 999; i++) {
        tasks.push(i);
    }

    // Process tasks with limited concurrency
    for (let i = 0; i < tasks.length; i += THREAD_COUNT) {
        const batch = tasks.slice(i, Math.min(i + THREAD_COUNT, tasks.length));
        const batchPromises = batch.map(num => runCollatzInWorker(num));
        const results = await Promise.all(batchPromises);
        collatzList.push(...results);
    }

    const endTime = Date.now();
    const elapsedTime = endTime - startTime;

    new CollatzInfoPrinter().prettyPrintCollatzList(collatzList);

    console.log(`the elapsed calculation time is ${elapsedTime.toString().padStart(8)} ms`);
}

main().catch(console.error);
